prometheus/collector.go

```diff
@@ -0,0 +1,156 @@
+package prometheus
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	exporter "github.com/kawamuray/prometheus-kafka-consumer-group-exporter"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+var (
+	partitionOffsetMetricsDesc = prometheus.NewDesc(
+		"kafka_broker_consumer_group_current_offset",
+		"Current consumed offset of a topic/partition",
+		[]string{"group_id", "consumer_address", "client_id", "topic", "partition"},
+		nil)
+	partitionLagMetricsDesc = prometheus.NewDesc(
+		"kafka_broker_consumer_group_offset_lag",
+		"Offset lag of a topic/partition",
+		[]string{"group_id", "consumer_address", "client_id", "topic", "partition"},
+		nil)
+)
+
+// PartitionInfoCollector is a Kafka prometheus.Collector. It uses a
+// exporter.PartitionInfoClient for the actual querying of Kafka.
+//
+// To speed up collection each consumer group is collected
+// concurrently.
+type PartitionInfoCollector struct {
+	groupListErrors     prometheus.Counter
+	groupDescribeErrors *prometheus.CounterVec
+
+	client exporter.PartitionInfoClient
+
+	ctx         context.Context
+	execTimeout time.Duration
+}
+
+// NewPartitionInfoCollector returns a prometheus.Collector that queries Kafka
+// using client. concurrency sets an upper limit on the number concurrent Kafka
+// concumer group queries running.
+func NewPartitionInfoCollector(client exporter.PartitionInfoClient, ctx context.Context, execTimeout time.Duration) *PartitionInfoCollector {
+	return &PartitionInfoCollector{
+		prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "kafka_broker_consumer_group_list_errors",
+			Help: "Number of Kafka scraping errors.",
+		}),
+		prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "kafka_broker_consumer_group_describe_errors",
+			Help: "Number of Kafka scraping errors.",
```

Comment:
That's a good question and a valid point! I suggest we push that forward in the future and fix that in a separate pull request (if needed). If you want, we can create an issue for it later.